### ISSUE TITLE
[Classic TBC] Fails to match "Attune" addon to CurseForge despite installing it via Ajour

### ISSUE THREAD

**Issue by Bananaman** [ajour/ajour#657]
## Update: It seems like I've discovered a bug in the official Curse Fingerprinting/Hashing algorithm. They generate 1 hash on their backend, and another hash using their official "fingerprinting algorithm". Both CurseForge (official app) and Ajour have this bug. This makes me think that the CurseForge *website backend* is using a slightly different algorithm for calculating the hash. If Ajour can figure out the difference (Attune-BCC.toc  is very small) and manage to generate the expected hash, you could be the first updater app with a bugfix for this CurseForge bug.

**Describe the bug**

Installing "Attune" from CurseForge via Ajour, and then re-starting Ajour (or pressing Refresh), will make the "Source" column say "Unknown" and auto-updates fail. I have a bunch of other addons and those all work. Yeah, I suspect that it may be something in Attune's toc file, but Ajour could probably still do a better job matching the addon instead of orphaning it. And that would enhance compatibility with all addons.

**To Reproduce**
1. Search Ajour's "Curse" catalog for "Attune".
2. Install "Attune" for TBC.
3. Restart Ajour or press Refresh.
4. See the "Source" column in Ajour's GUI, it will say "Unknown", and "Latest Version: -".

**Expected behavior**
Since it's a clearly named addon that came directly from Curse, it should be possible for Ajour to detect it and link it to its official CurseForge page.

**Software involved**
Please complete the following information:
 - OS: Windows 10
 - Ajour version: 1.2.0
 - Addon name: "Attune"
 - Flavor: "Classic TBC"

**Comment by Bananaman** [comment]
I went digging into the source code:

- https://github.com/ajour/ajour/blob/master/crates/core/src/parse.rs
- https://github.com/ajour/ajour/blob/master/crates/core/src/repository/backend/curse.rs

I can see that the process is something like this:

1. Download a list of fingerprints from wowup hub.
2. Scan local folders for addons matching those fingerprints. (The fingerprint seems to involve TOC tags like title, website URL and changelog URL?)
3. The Attune addons seems to get an incorrect fingerprint calculation on-disk by BOTH Ajour AND the official CurseForge App/Client. (Because when I tried it in the CurseForge Client, Attune always INSTANTLY says "Status: Modified" after a clean reinstallation via their official app.)

So it seems like in this case, it's a situation of an addon that is getting a wrong fingerprint on CurseForge's backend.

Any Ajour developers with experience in this fingerprinting may be able to figure out exactly what is wrong.

You can also join the Attune discord (see link at https://www.curseforge.com/wow/addons/attune) and then go to this message where we had some discussion about possible reasons why it's not matching properly: https://discord.com/channels/642161133477953546/790747167165251625/853767640966627348

**In summary:**

- It seems like CurseForge is generating fingerprint "X" for the Attune addon on their backend, but when CurseForge's official app OR YOURS (Ajour) tries to fingerprint the local files, you generate a DIFFERENT fingerprint, and this causes a mismatch.
- The issue exists in both CurseForge's app and in Ajour.
- In CurseForge's app, the "Status: Modified" means that auto-updates don't work and that the user has to manually force-update the addon.
- In Ajour's app, it simply can't even match "Attune" to Curse at all.

**Comment by Bananaman** [comment]
Forgot to mention: I looked in `%AppData%\ajour\cache` in the `fingerprints.yml` file. These are the details in there for Attune-230 (TBC):

```
  - title: Attune
    hash: 3976379423
    modified:
      secs_since_epoch: 1623627673
      nanos_since_epoch: 967510700
```

I didn't understand the fingerprinting code enough to know if this info helps, but it seems like it could be useful.

**Comment by casperstorm** [comment]
Thanks for your report, it is currently a known issue and a limitation of the API. Here is a comment from @tarkah on our Discord regarding it:

> The maintainer of that addon [Attune] needs to explicitly release the addon for each flavor. Right now he is releasing the single addon, but tagging it for both flavors. Unfortunately with the way the API works, it just shows up under the wow_burning_crusade  flavor. This is the same issue as when authors were posting an addon and tagging it as retail and classic. They need to publish it separately under each flavor. It'll be no different here. Expect this to be an issue for many more addons as TBC gets released.
this is a limitation not of Curseforge either, it really comes down to what interface the addon is marked as compatible with. Wow doesn't support an addon with "multiple" interface versions. Classic and TBC and Retail all have different interface versions. So it needs to be published separately with the correct Interface Version for each flavor it's actually compatible with as best practice (of course you can load out of date addons, but its not really the right way to go about it)

Let's for now keep this issue open to track it, and discuss potential solutions.

**Comment by Bananaman** [comment]
@casperstorm Ahhh, so basically this is the issue:

- Attune's author released 1 flavor (a zip file marked as retail on Curse, which is also including a BCC TOC).
- This makes makes Curse generate a fingerprint for the RETAIL "Attune.toc" file. But NO FINGERPRINT for the "Attune-BCC.toc" for the Classic TBC version and NO FINGERPRINT for "Attune-Classic" for the Classic Era version.
- So Curse says "Alright, Attune's (retail) fingerprint is: XYZ".
- But CurseForge's app AND Ajour's app instead fingerprint the "Attune-BCC.toc" file and will therefore generate a fingerprint that doesn't exist on Curse.

Alright how about this solution (it would make Ajour better than CurseForge's official app at matching):

- Calculate TWO local fingerprints: The current flavor, and then the retail .toc file's fingerprint. (Attune.toc or Attune-MainLine.toc)
- Do a lookup for the flavor fingerprint at curse.
- If that fails, do a lookup for the retail fingerprint at curse.

This matching would improve Ajour (since I bet that several other addons only release 1 flavor), but it would require a bit of extra code to make it do two fingerprints. :/ Thoughts?

**Comment by Bananaman** [comment]
By the way, CurseForge handles the Attune "single release" problem by always remembering the original source addon ID, since it only has itself as source. But it still gets confused by it (says "Status: Modified", since the CurseForge app is clearly calculating the TBC TOC fingerprint but expecting the Classic fingerprint from the backend release).

Ajour doesn't handle it at all, and becomes orphaned, since it can't match things back to the original installation source. Therefore the improved fingerprinting idea above is pretty much necessary in Ajour, to fix this and all other addons that use multi-version/single-zip releases.

**Comment by Bananaman** [comment]
Alright I had a long chat with the Ajour guys on Discord, and they said that Curse's old and new owners are aware of this issue (that uploading 1 zip file containing TOCs for multiple game versions = fingerprinting breaks). The issue is a few months old, since TBC classic added the "multiple TOC file per zip" system. The Curse website was not built to support multiple game versions per zip.

They say that it's on Curse's "todo" list to "fix someday". But for now, there is no fix. Addon authors must manually upload their addon zip file marked as separate game versions (classic, tbc, retail, etc). Until Curse fixes it.

Ajour *could* still do the workaround I described above, since it could take a very long time for Curse to fix their website (several months is likely). So feel free to consider the workaround if it's not too had code-wise. :-)

**Comment by casperstorm** [comment]
Since yesterday (https://github.com/ajour/ajour/commit/93cadf1c461c2e7b267f75d60baba242d1fabf29) we should have had multi toc support. It has also been released in 1.2.2.

@Bananaman, I'll close this issue now but please reopen if you run into any issues with any multi toc addon.
